Add support for additional flags

mjourdan opened the pull request:
## 🗣 Description ##

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

Wanted to automatically move the infected files to a quarantine folder. Also, allow some tuning.

This PR resolves #51 and #50.

## 🧪 Testing ##

I ran the role against a Debian 11 target host.

```yaml
- hosts: all
  become: yes
  become_method: sudo
  vars:
    clamav_scan_move: true
  roles:
    - clamav
```

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All relevant repo and/or project documentation has been updated to reflect the changes in this PR.
- [x] Tests have been added and/or modified to cover the changes in this PR.
- [x] All new and existing tests pass.

Commit:
Turn cron job from simple file to template

Commit:
Add support for quarantine and extra flags

Fixes #51

Commit:
replace exclude dir strings with regular expressions

fixes #50

Commit:
add quarantine dir to the list of exclusions

dav3r commented on README.md:
```diff
@@ -19,6 +19,14 @@ None.
 | Variable | Description | Default | Required |
 |----------|-------------|---------|----------|
 | clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
+| clamav_scan_exclude_directories | directory trees to exclude from scan operations         | { /dev, /sys, /proc } | No |
+| clamav_scan_move                | wether to move infected files to quarantine folder      | false | No |
+| clamav_scan_copy                | wether to copy infected files to "quarantine" folder    | false | No |
+| clamav_scan_quarantine_dir      | where to store infected files                           | '/var/spool/clamav' | No |
+| clamav_scan_quarantine_owner    | permissions on quarantine directory                     | 'root' | No |
+| clamav_scan_quarantine_group    | permissions on quarantine directory                     | 'root' | No |
+| clamav_scan_quarantine_mode     | permissions on quarantine directory                     | '0750' | No |
+| clamav_scan_extra_flags         | additional flags (see clamscan man page for reference)  | [] | No |
```
Alphabetize, clean up wording, and fix typos/formatting:
```suggestion
| clamav_scan_copy | Whether to copy infected files to quarantine folder. | false | No |
| clamav_scan_exclude_directories | Directory trees to exclude from scan operations. | `{ /dev, /proc, /sys, /var/spool/clamav }` | No |
| clamav_scan_extra_flags | Additional flags to pass to clamscan (see clamscan man page for reference).  | `[]` | No |
| clamav_scan_move | Whether to move infected files to a quarantine directory. | false | No |
| clamav_scan_quarantine_dir | Directory to store infected files. | `/var/spool/clamav` | No |
| clamav_scan_quarantine_group | Group owner to apply to quarantine directory. | `root` | No |
| clamav_scan_quarantine_mode | Permissions to apply to quarantine directory. | `0750` | No |
| clamav_scan_quarantine_owner | Owner to apply to quarantine directory. | `root` | No |
```

dav3r commented on defaults/main.yml:
```diff
@@ -1 +1,13 @@
 clamav_cron_frequency: weekly
+clamav_scan_exclude_directories:
+  - ^/dev/
+  - ^/sys/
+  - ^/proc/
+  - ^/var/spool/clamav
+clamav_scan_move: false
+clamav_scan_copy: false
+clamav_scan_quarantine_dir: '/var/spool/clamav'
+clamav_scan_quarantine_owner: root
+clamav_scan_quarantine_group: root
+clamav_scan_quarantine_mode: '0750'
+clamav_scan_extra_flags: []
```
We prefer to alphabetize wherever possible:
```suggestion
clamav_scan_copy: false
clamav_scan_exclude_directories:
  - ^/dev/
  - ^/proc/
  - ^/sys/
  - ^/var/spool/clamav
clamav_scan_extra_flags: []
clamav_scan_move: false
clamav_scan_quarantine_dir: '/var/spool/clamav'
clamav_scan_quarantine_group: root
clamav_scan_quarantine_mode: '0750'
clamav_scan_quarantine_owner: root
```

dav3r commented on defaults/main.yml:
```diff
@@ -1 +1,13 @@
 clamav_cron_frequency: weekly
+clamav_scan_exclude_directories:
+  - ^/dev/
+  - ^/sys/
+  - ^/proc/
+  - ^/var/spool/clamav
```
Are these `"^"` required here?  If not, we should remove them:
```suggestion
  - /dev/
  - /proc/
  - /sys/
  - /var/spool/clamav
```

mjourdan replied:
Actually these are required, as reported in #50 and documented in the [clamscan man page](https://linux.die.net/man/1/clamscan).

dav3r replied:
Thanks- I had not seen #50. 👍

dav3r commented on tasks/main.yml:
```diff
@@ -35,6 +35,15 @@
 
+- name: Ensure that quarantine folder exists
+  ansible.builtin.file:
+    path: "{{ clamav_scan_quarantine_dir }}"
+    state: directory
+    owner: "{{ clamav_scan_quarantine_owner }}"
+    group: "{{ clamav_scan_quarantine_group }}"
+    mode: "{{ clamav_scan_quarantine_mode }}"
+  when: clamav_scan_move or clamav_scan_copy | bool
```
More alphabetizing:
```suggestion
    group: "{{ clamav_scan_quarantine_group }}"
    mode: "{{ clamav_scan_quarantine_mode }}"
    owner: "{{ clamav_scan_quarantine_owner }}"
    path: "{{ clamav_scan_quarantine_dir }}"
    state: directory
  when: clamav_scan_copy or clamav_scan_move | bool
```

mjourdan replied:
I have doubts with this one. I would rather keep the order here. I find it is easier to read because:
- it makes more sense to have the module target as the first argument
- the module arguments are kept in the same order as the [documentation](https://docs.ansible.com/ansible/latest/collections/ansible/builtin/file_module.html), which is allegedly the most common

dav3r replied:
Although I understand where you are coming from, alphabetizing module arguments is another one of our standards.  We find it simpler to parse and update lists of things when they are in alphabetical order.

mjourdan replied:
Alright, I alphabetized both module arguments in the tasks and command flags in the script. I believe everything is ok now. Thanks for your review @dav3r!

dav3r replied:
Thank you! 👍

dav3r commented on templates/virus_scan.sh:
```diff
@@ -1,18 +1,28 @@
-# Scan the entire file system (modulo the /dev, /sys, and /proc trees)
+# Scan the entire file system (modulo excluded trees)
 # and write to the log
 clamscan --infected --recursive \
```
This command should be updated so that all of the clamscan flags are listed alphabetically.

mjourdan replied:
To me it just make more sense to have the options specifically handed by the role at the beginning, and the vague set of additional and facultative options at the end. Hence the order both in the script and in the defaults. No big deal though.

dav3r commented on templates/virus_scan.sh:
```diff
@@ -1,18 +1,28 @@
 #!/bin/bash
```
This file should be renamed to `templates/virus_scan.tpl.sh` to match [our (undocumented) standard](https://github.com/cisagov/cool-assessment-terraform/tree/develop/cloud-init).

dav3r reviewed (changes requested):
@mjourdan Thanks very much for contributing to this repo!  Please take a look at my suggestions and let me know if you have any questions or concerns.

mjourdan committed:
Update README.md

Alphabetize, clean up wording, and fix typos/formatting

Co-authored-by: dav3r <daver@geekpad.com>

mjourdan committed:
Update defaults/main.yml

Alphabetize

Co-authored-by: dav3r <daver@geekpad.com>

Commit:
Rename template to match cisagov standard

mjourdan committed:
Update tasks/main.yml

Co-authored-by: dav3r <daver@geekpad.com>

Commit:
alphabetize clamscan flags

dav3r reviewed (approved):
Looks great to me now - thank you @mjourdan for contributing!

@jsf9k, @mcdonnnj, or @felddy - can you please give this PR a once-over?

mcdonnnj commented on README.md:
```diff
@@ -19,6 +19,14 @@ None.
 | Variable | Description | Default | Required |
 |----------|-------------|---------|----------|
 | clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
+| clamav_scan_copy | Whether to copy infected files to quarantine folder. | false | No |
```
```suggestion
| clamav_scan_copy | Whether to copy infected files to quarantine folder. | `false` | No |
```

mcdonnnj commented on README.md:
```diff
@@ -19,6 +19,14 @@ None.
+| clamav_scan_copy | Whether to copy infected files to quarantine folder. | false | No |
+| clamav_scan_exclude_directories | Directory trees to exclude from scan operations. | `{ /dev, /proc, /sys, /var/spool/clamav }` | No |
+| clamav_scan_extra_flags | Additional flags to pass to clamscan (see clamscan man page for reference).  | `[]` | No |
+| clamav_scan_move | Whether to move infected files to a quarantine directory. | false | No |
```
```suggestion
| clamav_scan_move | Whether to move infected files to a quarantine directory. | `false` | No |
```

mcdonnnj commented on README.md:
```diff
@@ -19,6 +19,14 @@ None.
+| clamav_scan_exclude_directories | Directory trees to exclude from scan operations. | `{ /dev, /proc, /sys, /var/spool/clamav }` | No |
+| clamav_scan_extra_flags | Additional flags to pass to clamscan (see clamscan man page for reference).  | `[]` | No |
+| clamav_scan_move | Whether to move infected files to a quarantine directory. | false | No |
+| clamav_scan_quarantine_dir | Directory to store infected files. | `/var/spool/clamav` | No |
```
Minor thing but I prefer to avoid abbreviations when feasible. I believe I've appropriately suggested all other appearances of this variable but apologies if I've missed one.

```suggestion
| clamav_scan_quarantine_directory | Directory to store infected files. | `/var/spool/clamav` | No |
```

mcdonnnj commented on defaults/main.yml:
```diff
@@ -1 +1,13 @@
+  - ^/var/spool/clamav
+clamav_scan_extra_flags: []
+clamav_scan_move: false
+clamav_scan_quarantine_dir: '/var/spool/clamav'
```
```suggestion
clamav_scan_quarantine_directory: '/var/spool/clamav'
```

jsf9k replied:
The single quotes are not necessary here and can be removed.
```suggestion
clamav_scan_quarantine_dir: /var/spool/clamav
```

mcdonnnj commented on tasks/main.yml:
```diff
@@ -35,6 +35,15 @@
+    group: "{{ clamav_scan_quarantine_group }}"
+    mode: "{{ clamav_scan_quarantine_mode }}"
+    owner: "{{ clamav_scan_quarantine_owner }}"
+    path: "{{ clamav_scan_quarantine_dir }}"
```
```suggestion
    path: "{{ clamav_scan_quarantine_directory }}"
```

mcdonnnj commented on templates/virus_scan.tpl.sh:
```diff
@@ -0,0 +1,33 @@
+# and write to the log
+clamscan \
+{% if clamav_scan_copy %}
+  --copy={{ clamav_scan_quarantine_dir }} \
```
```suggestion
  --copy={{ clamav_scan_quarantine_directory }} \
```

mcdonnnj commented on templates/virus_scan.tpl.sh:
```diff
@@ -0,0 +1,33 @@
+  --infected \
+  --log=${LAST_SCAN_LOG_FILENAME} \
+{% if clamav_scan_move %}
+  --move={{ clamav_scan_quarantine_dir }} \
```
```suggestion
  --move={{ clamav_scan_quarantine_directory }} \
```

mcdonnnj reviewed (changes requested):
Thanks for your contribution! This all looks solid. I have some minor changes I'd like to see below.

jsf9k commented on README.md:
```diff
@@ -19,6 +19,14 @@ None.
 |----------|-------------|---------|----------|
 | clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
+| clamav_scan_copy | Whether to copy infected files to quarantine folder. | false | No |
+| clamav_scan_exclude_directories | Directory trees to exclude from scan operations. | `{ /dev, /proc, /sys, /var/spool/clamav }` | No |
```
This is `list`, not a `dict`:
```suggestion
| clamav_scan_exclude_directories | A list of regexes matching irectory trees that are to be excluded from scan operations. | `[^/dev, ^/proc, ^/sys, ^/var/spool/clamav]` | No |
```

I also updated the default values to match those in `defaults/main.yml` and tweaked the variable description to mention that the list entries are regexes.

jsf9k commented on defaults/main.yml:
```diff
@@ -1 +1,13 @@
+clamav_scan_move: false
+clamav_scan_quarantine_dir: '/var/spool/clamav'
+clamav_scan_quarantine_group: root
+clamav_scan_quarantine_mode: '0750'
```
The single quotes are not necessary here and can be removed.
```suggestion
clamav_scan_quarantine_mode: 0750
```

jsf9k reviewed (changes requested):
I'd like to see some additional tests that cover the new options you have added.  It may make sense to create some new `molecule` scenarios for this, as is done, e.g., in cisagov/ansible-role-python#46.

mjourdan committed:
Apply suggestions from code review

avoid abbreviation

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

mjourdan committed:
Update README.md

replace dict with list

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

mjourdan committed:
Update defaults/main.yml

remove unnecessary quotes on file mode

Co-authored-by: Shane Frasier <maverick@maverickdolphin.com>

Commit:
remove unnecessary quotes on quarantine directory

mjourdan committed:
Apply suggestions from code review

Add backticks to default values in readme

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

ClementJ35 committed:
[tests] First commit of tree folder

ClementJ35 committed:
[tests] Second commit of the new folder tree

ClementJ35 committed:
[tests] Write first test

ClementJ35 committed:
[test] add comment and fix syntax

ClementJ35 committed:
[tests] Fix lint & add the scenarios to the matrix

ClementJ35 committed:
[tests] fix scenario names

ClementJ35 committed:
[tests] Add needed switch to perform actions

ClementJ35 committed:
[template] Rename to be more explicit about content

ClementJ35 committed:
[tests] extract shell content as string instead of binary

ClementJ35 committed:
[tests] Bitify instead of Stringify

ClementJ35 committed:
[tests] Exclude bandit because of assert

ClementJ35 committed:
[tests] Fix typo in bandit exclusion & fix linting

ClementJ35 committed:
[tests] real fix of linting

ClementJ35 committed:
[tests] Make linters happy

ClementJ35 committed:
[tests] duplicate 0c66f8f8c33b585da3bad7d1a7b6714f8f2aa396 for the new tests

ClementJ35 committed:
[tests] handle line break in string declaration

ClementJ35 committed:
[tests] Fix dummy switches order in content_test

ClementJ35 commented:
Hi @jsf9k,
I did the tests as you did for modify_config. It works nice (see https://github.com/mjourdan/ansible-role-clamav/actions/runs/3250333690 ) 
Here, it can't run because of merge conflict.
What is your policy ?
Do we merge / rebase / do nothing ?

jsf9k commented:
> Hi @jsf9k, I did the tests as you did for modify_config. It works nice (see https://github.com/mjourdan/ansible-role-clamav/actions/runs/3250333690 ) Here, it can't run because of merge conflict. What is your policy ? Do we merge / rebase / do nothing ?

Thanks @ClementJ35 and @mjourdan!  I rebased the PR; once the tests pass I'll re-review.

jsf9k commented:
The build is failing because of [an issue in `molecule`](https://github.com/ansible-community/molecule/pull/3686).  Hopefully that issue is resolved upstream soon.

ClementJ35 commented:
Hi @jsf9k, thanks for rebasing.
It seems your PR as been merged : https://github.com/ansible-community/molecule/pull/3687
I'll keep an eye here to see if it's passing all tests.

jsf9k commented:
> Hi @jsf9k, thanks for rebasing. It seems your PR as been merged : [ansible-community/molecule#3687](https://github.com/ansible-community/molecule/pull/3687) I'll keep an eye here to see if it's passing all tests.

Yep, once they release a new version of Molecule I can re-run the tests and they should pass.

jsf9k committed:
Update test code to match long CLI option in virus_scan script

jsf9k committed:
Modify test to make it less brittle

jsf9k reviewed (approved):
Great work @mjourdan!  Thank you for the contribution!

jsf9k commented:
@dav3r and @mcdonnnj, please re-review this pull request.

dav3r reviewed (approved):
Looks great - strong work! 🚀

mcdonnnj commented on README.md:
```diff
@@ -22,6 +22,14 @@ None.
 | clamav_cron_frequency | The frequency of ClamAV scanning.  Must be one of: `hourly`, `daily`, `weekly`, or `monthly`. | `weekly` | No |
 | clamav_freshclam_configuration | A dictionary of values to set in the freshclam configuration file. | `{}` | No |
+| clamav_scan_copy | Whether to copy infected files to quarantine folder. | `false` | No |
+| clamav_scan_exclude_directories | A list of regexes matching irectory trees that are to be excluded from scan operations. | `[^/dev, ^/proc, ^/sys, ^/var/spool/clamav]` | No |
```
```suggestion
| clamav_scan_exclude_directories | A list of regexes matching directory trees that are to be excluded from scan operations. | `[^/dev, ^/proc, ^/sys, ^/var/spool/clamav]` | No |
```

mcdonnnj commented on molecule/additional_flags/INSTALL.rst:
```diff
@@ -0,0 +1,22 @@
+*******
```
Maybe nit-picky but should this be a symlink like it is for `molecule/modify_config`?

jsf9k replied:
Please see commit 9713cb1.

jsf9k replied:
Please see commit 2e1b063.

mcdonnnj commented on molecule/additional_flags/prepare.yml:
```diff
@@ -0,0 +1,50 @@
+---
```
Maybe nit-picky but should this be a symlink like it is for `molecule/modify_config`?

jsf9k replied:
Please see commit 9713cb1.

jsf9k replied:
Please see commit 2e1b063.

mcdonnnj commented on molecule/additional_flags/python.yml:
```diff
@@ -0,0 +1,8 @@
+---
```
Maybe nit-picky but should this be a symlink like it is for `molecule/modify_config`?

jsf9k replied:
Please see commit 9713cb1.

mcdonnnj commented on molecule/additional_flags/requirements.yml:
```diff
@@ -0,0 +1,9 @@
+---
```
Maybe nit-picky but should this be a symlink like it is for `molecule/modify_config`?

jsf9k replied:
Please see commit 9713cb1.

mcdonnnj commented on molecule/additional_flags/upgrade.yml:
```diff
@@ -0,0 +1,7 @@
+---
```
Maybe nit-picky but should this be a symlink like it is for `molecule/modify_config`?

jsf9k replied:
Please see commit 9713cb1.

mcdonnnj commented on molecule/additional_flags/molecule.yml:
```diff
@@ -0,0 +1 @@
+molecule-with-systemd.yml
```
Should this be set up like the one for `molecule/modify_config`?

jsf9k replied:
Please see commit 9713cb1.

mcdonnnj commented on molecule/additional_flags/molecule-no-systemd.yml:
```diff
@@ -0,0 +1,74 @@
+---
```
If `molecule.yml` is changed this file would no longer be needed.

jsf9k replied:
Please see commit 9713cb1.

mcdonnnj commented on molecule/additional_flags/molecule-with-systemd.yml:
```diff
@@ -0,0 +1,124 @@
+---
```
If `molecule.yml` is changed this file would no longer be needed.

jsf9k replied:
Please see commit 9713cb1.

mcdonnnj reviewed (approved):
Thanks for the continued work in getting this implemented. I'm suggesting that the `additional_flags` scenario's files more closely mirror the form of the `modify_config` scenario so that we are consistent. Other than that all I saw was a typo.

jsf9k committed:
Fix typo

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k committed:
Organize additional_flags scenario similarly to modify_config

This entailed making several files in molecule/additional_flags
symlinks instead of hard copies.  I also broke the tests that were in
default/tests/test_default.py into two files so I could reuse most of
them in molecule/additional_flags but still create a slightly
different version of others specific to the scenario.

Co-authored-by: Nicholas McDonnell <50747025+mcdonnnj@users.noreply.github.com>

dav3r commented on molecule/additional_flags/tests/test_files_and_dirs_additional_flags.py:
```diff
@@ -0,0 +1,26 @@
+"""Module containing the tests for the default scenario."""
```
Update scenario name:
```suggestion
"""Module containing the tests for the additional_flags scenario."""
```

dav3r reviewed (approved):
Still good, aside from one missed docstring update. 🚀

jsf9k committed:
Update docstring to indicate the correct scenario

Co-authored-by: dav3r <daver@geekpad.com>

mcdonnnj commented on molecule/additional_flags/tests/test_additional_flags.py:
```diff
@@ -0,0 +1,27 @@
+"""Module containing the tests for the additional flags scenario."""
```
```suggestion
"""Module containing the tests for the additional_flags scenario."""
```

mcdonnnj reviewed (approved):
Just a small docstring fix from me.

jsf9k committed:
Add an underscore in a scenario name in a docstring

Co-authored-by: Nick <50747025+mcdonnnj@users.noreply.github.com>

jsf9k commented:
@mjourdan and @ClementJ35, thanks for the contribution and for sticking with us.  This one is finally merged!

mjourdan commented:
Thank you @jsf9k @dav3r @mcdonnnj @ClementJ35  \o/
